cocos/platform/android/java/src/org/cocos2dx/lib/Cocos2dxDownloader.java

```diff
@@ -1,319 +1,300 @@
-    private HashMap _taskMap = new HashMap();
-    private Queue<Runnable> _taskQueue = new LinkedList<Runnable>();
-
-    void onProgress(final int id, final long downloadBytes, final long downloadNow, final long downloadTotal) {
-        DownloadTask task = (DownloadTask)_taskMap.get(id);
-        if (null != task) {
-            task.bytesReceived = downloadBytes;
-            task.totalBytesReceived = downloadNow;
-            task.totalBytesExpected = downloadTotal;
-        }
-        Cocos2dxHelper.runOnGLThread(new Runnable() {
-            @Override
-            public void run() {
-                nativeOnProgress(_id, id, downloadBytes, downloadNow, downloadTotal);
-            }
-        });
+    public void createTask(final int taskID, final String url, final String path) {
+	Runnable taskRunnable = new Runnable() {
+	    @Override
+	    public void run() {
+		RequestHandle handle;
+
+		if (0 == path.length()) {
+		    // data task
+		    DataTaskHandler handler = new DataTaskHandler(taskID);
+		    handle = _httpClient.get(Cocos2dxHelper.getActivity(), url, handler);
+		} else {
+		    // file task
+		    File tempFile = new File(path + _tempFileNameSufix);
+		    if (tempFile.isDirectory()) {
+			final String errStr = "Can't create DownloadTask for " + url + ". Unable to create temp file: " + tempFile;
+			Cocos2dxHelper.runOnGLThread(new Runnable() {
+			    @Override
+			    public void run() {
+				nativeOnFinish(_id, taskID, 0, errStr, null);
+			    }
+			});
+			return;
+		    }
+
+		    File parent = tempFile.getParentFile();
+		    if (!parent.isDirectory() && !parent.mkdirs()) {
+			final String errStr = "Can't create DownloadTask for " + url + ". Unable to create temp file directory: " + parent;
+			Cocos2dxHelper.runOnGLThread(new Runnable() {
+			    @Override
+			    public void run() {
+				nativeOnFinish(_id, taskID, 0, errStr, null);
+			    }
+			});
+			return;
+		    }
+
+		    File finalFile = new File(path);
+		    if (finalFile.isDirectory()) {
+			final String errStr = "Can't create DownloadTask for " + url + ". Unable to create file: " + finalFile;
+			Cocos2dxHelper.runOnGLThread(new Runnable() {
+			    @Override
+			    public void run() {
+				nativeOnFinish(_id, taskID, 0, errStr, null);
+			    }
+			});
+			return;
+		    }
+
+		    FileTaskHandler handler = new FileTaskHandler(taskID, tempFile, finalFile);
+		    Header[] headers = null;
+		    long fileLen = tempFile.length();
+		    if (fileLen > 0) {
+			// continue download
+			headers = new Header[1];
+			headers[0] = new BasicHeader("Range", "bytes=" + fileLen + "-");
+		    }
+		    handle = _httpClient.get(Cocos2dxHelper.getActivity(), url, headers, null, handler);
+		    //task.handle = downloader._httpClient.get(url, task.handler);
+		}
+
+		taskHandles.put(taskID, handle);
+	    }
+	};
+	if (_taskQueue.size() < _countOfMaxProcessingTasks) {
+	    Cocos2dxHelper.getActivity().runOnUiThread(taskRunnable);
```

Comment:
Litterally this line breaks my logic that fixed the out of memory problem in https://github.com/cocos2d/cocos2d-x/pull/14530, making the _taskQueue totally useless. You have to push a null into the queue if you are to execute the task immediately otherwise the _taskQueue is forever empty